lib/galaxy/jobs/__init__.py

```diff
@@ -1072,7 +1072,15 @@ def change_state(self, state, info=False, flush=True, job=None):
             dataset = dataset_assoc.dataset
             if not job_supplied:
                 self.sa_session.refresh(dataset)
-            dataset.raw_set_dataset_state(state)
+            state_changed = dataset.raw_set_dataset_state(state)
+            if state_changed:
+                # Arguably a hack to get state changes to appear in the history panel because
+                # the history panel polls on hda.update_time and ignores hda.dataset.update_time.
```

Comment:
I'm a bit confused, aren't you updating `hda.dataset.update_time` below? `dataset` should be a `DatasetInstance` instance after line 1072. And otherwise the changes to `lib/galaxy/model/__init__.py` wouldn't be needed.

Reply:
My intention was to update hda.update_time explicitly now - hda.dataset.update_time was already being set. I think I understand the confusion. If I changed the name “hda” in the comment to “dataset_instance” and I changed the local variable name in the code from dataset to “dataset_instance” would this clarify things better? Galaxy is pretty consistently inconsistent about naming these things - which I’ve definitely contributed to.

Reply:
OK, so `dataset_assoc` is an `JobToOutputDatasetAssociation` or `JobToOutputLibraryDatasetAssociation` object and `dataset` is a `HistoryDatasetAssociation` or `LibraryDatasetDatasetAssociation` object. Both of these last 2 classes derive from `DatasetInstance`.
I think my main confusion was that I supposed that `dataset_assoc` was already a `HistoryDatasetAssociation` and `dataset` was a `Dataset` instance (although I originally wrote `DatasetInstance` in the comment above). So the changes are fine, variable renames can wait, but thanks for the offer!